File: DP/494. Target Sum/494. Target Sum.py
class Solution(object):
    def findTargetSumWays(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        Sum = sum(nums)
        if (abs(target) > Sum): return 0
        
        N = len(nums)
        dp=[[0 for _ in range(2*Sum+1)] for _ in range(N)]
        dp[0][nums[0]+Sum] +=1
        dp[0][-nums[0]+Sum] +=1
        
        for i in range(1,N):
            for s in range(-Sum, Sum+1):
                s1 = s+nums[i]
                s2 = s-nums[i]
                
                if (abs(s1) <= Sum): dp[i][s1+Sum] += dp[i-1][s+Sum]
                if (abs(s2) <= Sum): dp[i][s2+Sum] += dp[i-1][s+Sum]
                    
        return dp[N-1][target+Sum]

File: DP/494. Target Sum/494. Target Sum.java
class Solution {
    // dfs + prune => dp
    public int findTargetSumWays(int[] nums, int target) {
        int n = nums.length;
        int[] newnums = new int[n+1];
        for (int i = 0; i < n; i++) newnums[i+1] = nums[i];
        int sum = 0;
        for (int c : nums) sum += c;
        if (sum < target) return 0;
        
        int[][] dp = new int[n+1][2*sum+1];
        dp[0][sum] = 1;
        
        for (int i = 1; i < n+1; i++) {
            for (int j = -sum; j <= sum; j++) {
                if (j+sum+newnums[i] <= 2*sum) 
                    dp[i][j+sum+newnums[i]] += dp[i-1][j+sum];
                if (j+sum-newnums[i] >= 0)
                    dp[i][j+sum-newnums[i]] += dp[i-1][j+sum];
            }
        }
        return dp[n][target+sum];
    }
}
// dp[s+num[i]]: dpi-1[s]
// dp[s-num[i]]: dpi-1[s]
